azavea/raster-foundry
# Mosaic UI

## Overview

This does two main things:
- Refactors the color correction interface to use UI-Router states so that the back button works properly (#725).
- Adds a largely non-functional Mosaic scenes list UI, with draggable scene reordering.

### Checklist


### Demo

![mosaic](https://cloud.githubusercontent.com/assets/447977/21064477/f4d82214-be28-11e6-9043-06e40b37d5cb.png)

### Notes
A general note on architecture: In order to allow the URI to more closely track what is going on in the interface, and to avoid having really complex state logic in the `ProjectEditorController` to manage hiding/showing multiple different color correction and mosaic interfaces, I broke apart what was previously the `ProjectEditorController` into several components. `ProjectEditorController` (`editor.project`) is now an abstract state, which has two further abstract child states, `editor.project.mosaic` and `editor.project.color`. These two abstract states in turn have child states corresponding to their interface panes, which means that a there are currently the following concrete states: `editor.project.color.scenes`, `editor.project.color.adjust`, `editor.project.mosaic.scenes`. To add further mosaic views, we can add more states to this.

Also, the functionality that was previously the `colorCorrectPane` component has been moved; it now mostly lives in the `colorCorrectAdjust` component. There is a new `colorCorrectPane` component which coordinates between the `colorCorrectAdjust` component and the histogram, and provides functionality for the `Back` and `Reset` buttons.

There are some known issues / difficulties with this:
- The HTML nesting had to be restructured to get the nested ui-router states to work; this broke the styling on the scene lists and the map height determination. I've talked to @designmatty and will create a design issue so that he can take care of it.
- Passing state, such as selected scenes and map layers, back and forth correctly, cleanly, and efficiently between the various nested ui-router states turned out to be difficult. I eventually got a working solution by using two-way bindings to the abstract parent state (`ProjectEditController`), but this isn't ideal from a performance perspective. Also, in practice it acts like a global shared mutable state between the color correction and mosaic views, which isn't great. I'm not sure whether ui-router's forthcoming 1.0.0 release will help to alleviate some of these issues but this is an area that could use some fresh eyes; my hope is that I'm missing something easy and obvious.
- I am still getting a lot of 500 errors from the tile server even after rebasing off the latest develop. This is making it difficult to diagnose a bug which is that it seems to get confused about which layers are selected if you switch back and forth between the color correction scene list and adjustment panel and then try to color correct again.

## Testing Instructions

 * From `app-frontend`, run `npm start`; also start up a `./scripts/server` from inside the VM.
 * Verify that the color correction interface still works as expected (subject to the layout issues above).
 * Verify that you can switch to the Mosaic view by clicking on the Mosaic header and that it loads a scene list which you can grab by the left handles and reorder.
 * Verify that URLs update throughout this process and that the back button works properly.

Closes #725
